# HG changeset patch
# User Rik <rik@octave.org>
# Date 1523565050 25200
#      Thu Apr 12 13:30:50 2018 -0700
# Node ID c64f7a95b464242a94d2b67a929f6ee585a35f7f
# Parent  8a92b651838b204b3b28cf7a22500f48b1f1d25a
pie.m, pie3.m: Validate that data input is finite (bug #53480).

* pie.m, pie3.m: Use isfinite() to check input data X.
Add BIST tests for input validation.

diff --git a/scripts/plot/draw/pie.m b/scripts/plot/draw/pie.m
--- a/scripts/plot/draw/pie.m
+++ b/scripts/plot/draw/pie.m
@@ -55,16 +55,20 @@
 function h = pie (varargin)
 
   [hax, varargin, nargin] = __plt_get_axis_arg__ ("pie", varargin{:});
 
   if (nargin < 1)
     print_usage ();
   endif
 
+  if (! all (isfinite (varargin{1})))
+    error ("pie: all data in X must be finite"); 
+  endif
+
   oldfig = [];
   if (! isempty (hax))
     oldfig = get (0, "currentfigure");
   endif
   unwind_protect
     hax = newplot (hax);
     htmp = __pie__ ("pie", hax, varargin{:});
   unwind_protect_cleanup
@@ -95,8 +99,13 @@ endfunction
 
 
 %!demo
 %! clf;
 %! pie ([0.17, 0.34, 0.41], {"Cheddar", "Swiss", "Camembert"});
 %! colormap ([1,0,0;0,1,0;0,0,1;1,1,0;1,0,1;0,1,1]);
 %! axis ([-2,2,-2,2]);
 %! title ("pie() with missing slice");
+
+## Test input validation
+%!error pie ()
+%!error <all data in X must be finite> pie ([1 2 Inf])
+%!error <all data in X must be finite> pie ([1 2 NaN])
diff --git a/scripts/plot/draw/pie3.m b/scripts/plot/draw/pie3.m
--- a/scripts/plot/draw/pie3.m
+++ b/scripts/plot/draw/pie3.m
@@ -56,16 +56,20 @@
 function h = pie3 (varargin)
 
   [hax, varargin, nargin] = __plt_get_axis_arg__ ("pie3", varargin{:});
 
   if (nargin < 1)
     print_usage ();
   endif
 
+  if (! all (isfinite (varargin{1})))
+    error ("pie3: all data in X must be finite"); 
+  endif
+
   oldfig = [];
   if (! isempty (hax))
     oldfig = get (0, "currentfigure");
   endif
   unwind_protect
     hax = newplot (hax);
     htmp = __pie__ ("pie3", hax, varargin{:});
   unwind_protect_cleanup
@@ -95,8 +99,13 @@ endfunction
 %! title ("pie3() with labels");
 
 %!demo
 %! clf;
 %! pie3 ([0.17, 0.34, 0.41], {"Cheddar", "Swiss", "Camembert"});
 %! colormap ([1,0,0;0,1,0;0,0,1;1,1,0;1,0,1;0,1,1]);
 %! axis ([-2,2,-2,2]);
 %! title ("pie3() with missing slice");
+
+## Test input validation
+%!error pie3 ()
+%!error <all data in X must be finite> pie3 ([1 2 Inf])
+%!error <all data in X must be finite> pie3 ([1 2 NaN])
